### PR TITLE
Docs CI Rework for Versioned Deploy

### PR DESCRIPTION
**Pull request by stanbrub**
Added _docs-ci-v2.yml_ to deploy to directory deephaven-core-v2, while _docs-ci.yml_ remains unchanged.

- Pull request to for _main_
   - Build the docs
- Push to _main_ 
   - Build the docs
   - Upload artifacts to Github
   - Deploy docs to deephaven.io  (e.g. _~/deephaven-core-v2/main_)
- Push to _release/v*_
   - Set symlinks for _latest_ (pointing to newest release dir)
   - Build the docs
   - Upload artifacts to Github
   - Deploy docs to release dir (e.g. ~/deephaven-core-v2/release/v0.30.0)

New Directory structure
- Paired down to include mostly docs directories
- _deephaven-core-v2_ is what's new
```
├── deephaven-core
│   ├── client-api
│   ├── javadoc
│   └── pydoc
├── deephaven-core-v2
│   ├── main
│   │   ├── client-api
│   │   ├── javadoc
│   │   └── pydoc
│   ├── release
│   │   ├── v0.30.0
│   │   │   ├── client-api
│   │   │   ├── javadoc
│   │   │   └── pydoc
│   │   └── v0.31.0
│   │   │  ├── client-api
│   │   │  ├── javadoc
│   │   │  └── pydoc
│   └── symlinks
│   │   ├── latest -> ../release/v0.31.0
│   │   └── next -> ../main
├── deephaven.io
│   ├── barrage
│   │   └── docs
│   ├── core
│   │   ├── docs
│   │   └── groovy
│   ├── enterprise
│   │   ├── docs
│   │   ├── jackson
│   │   ├── powell
│   │   ├── services
│   │   ├── silverheels
│   └── grid
        ├── How-to
        └── docs
```

### CODE REVIEW

File: .github/workflows/docs-ci.yml
Comment (devinrsmith) [inline .github/workflows/docs-ci.yml:38]: I'm a bit scared about rsyncing w/ symlinks into a root directory. Not that I this is necessarily know if there is anything wrong w/ how it is expressed now, but I imagine things could go wrong w/ a mistype. Maybe we want an additional subdirectory? `deephaven-core/symlinks/`?

**Review by devinrsmith — COMMENTED** [review]
I did not audit the actual arguments to rsync, but we'll want other eyes to look in depth before we merge.

It would be good to see a `tree` of the proposed layout.

**Comment by dsmmcken** [comment]
> What do these rsync options do?
> `-rlptDvz --force --mkpath`

> The rsync options you provided do the following:
> 
> -r: This option makes rsync perform a recursive directory transfer.
> -l: This option makes rsync copy symbolic links as symbolic links.
> -p: This option preserves file permissions.
> -t: This option preserves modification times.
> -D: This option is equivalent to --devices --specials, preserving devices and special files.
> -v: This option increases verbosity. It causes rsync to give more information about what it's doing.
> -z: This option compresses the data as it is sent to the destination machine, which reduces the amount of data being transmitted.
> --force: This option forces deletion of directories even if not empty.
> --mkpath: This option is not a standard rsync option. It might be specific to a certain version or variant of rsync.
> Please note that I'm an AI programming assistant and my expertise is strictly limited to software development topics.

co-pilot seems to think --mkpath isn't an option, what does it do?

File: .github/workflows/docs-ci.yml
Comment (stanbrub) [reply .github/workflows/docs-ci.yml:10]: Using the ternary and the commensurate conditions in the steps using _version-path_ makes it much easier to test both main and release pushes from a fork branch and a test server.  You can just "**" the trigger and change between "next" and "release/v0.30.0" to get the same behavior.

File: .github/workflows/docs-ci.yml
Comment (stanbrub) [reply .github/workflows/docs-ci.yml:13]: Added needs for _symlink_ for all jobs.  There was a chance _latest_ could point to an unfinished release/v* directory.

File: .github/workflows/docs-ci.yml
Comment (stanbrub) [reply .github/workflows/docs-ci.yml:25]: I don't understand the "weird" part.  Variables are part of Workflows.  Are they unstable?  As for why I did it this way... as mentioned above, it's just easier to test.

File: .github/workflows/docs-ci.yml
Comment (stanbrub) [reply .github/workflows/docs-ci.yml]: Done

File: .github/workflows/docs-ci.yml
Comment (stanbrub) [reply .github/workflows/docs-ci.yml:41]: Rsyncing links to the _deephaven-core_ directory seems no different than rsyncing _javadoc_ or _pydoc_ dirs into the same directory.  How are links more dangerous than those overwriting directories?  (Also, there's no _--delete_ on this deploy.)

The other issue is that for backward compatibility, at least until the deephaven.io hrefs change, the_deephaven-core/javadoc_ et al dirs are replaced with symlinks in the _next_ directory.

By the way, you probably already know this, but the _deephaven-core/_ directory is a subdirectory under a user that contains only Docs CI deployed api docs.  We aren't affecting anything outside it.

**Comment by stanbrub** [comment]
> > What do these rsync options do?
> > `-rlptDvz --force --mkpath`
> 
> > The rsync options you provided do the following:
> > -r: This option makes rsync perform a recursive directory transfer.
> > -l: This option makes rsync copy symbolic links as symbolic links.
> > -p: This option preserves file permissions.
> > -t: This option preserves modification times.
> > -D: This option is equivalent to --devices --specials, preserving devices and special files.
> > -v: This option increases verbosity. It causes rsync to give more information about what it's doing.
> > -z: This option compresses the data as it is sent to the destination machine, which reduces the amount of data being transmitted.
> > --force: This option forces deletion of directories even if not empty.
> > --mkpath: This option is not a standard rsync option. It might be specific to a certain version or variant of rsync.
> > Please note that I'm an AI programming assistant and my expertise is strictly limited to software development topics.
> 
> co-pilot seems to think --mkpath isn't an option, what does it do?

In order to place versioned directories like _release/v0.30.0_, the _remote-path_ action attribute has a directory that doesn't exist.  _--mkpath_ creates it.  Without it we'd have to manipulate each set of docs locally to have the correct directory structure.  That would create a bunch of more steps in Docs CI.  (If you'd feel safer with it that way, I can do it.)

_--force_ is necessary to overwrite the existing javadoc, pydoc, and client-api directories with links.

File: .github/workflows/docs-ci.yml
Comment (devinrsmith) [reply .github/workflows/docs-ci.yml:25]: The "weird" part is that you are attaching specific semantics to a directory path name that could otherwise be any string - and to understand the `if` statement, now I have to go understand what `version-path` is. Additionally, if we are going to change `version-path` in the future, we know need to audit everything that uses it to understand what `version-path` _actually_ is. In this PR, `version-path` *isn't* just a directory name, but it is also being used for this dual-purpose of gating whether a step runs or not, and that's why it's "weird".

File: .github/workflows/docs-ci.yml
Comment (devinrsmith) [inline .github/workflows/docs-ci.yml:7]: I'm not sure I would universally change this to `**`. The convention we use in other workflows is to have a specific prefix. I think if you did `branches: [ 'main', 'docs/**', 'release/v*' ]`, that would be more in line w/ our current conventions (see `build-ci.yml`).

File: .github/workflows/docs-ci.yml
Comment (devinrsmith) [inline .github/workflows/docs-ci.yml:10]: I'm assuming this `|| 'release/v0.30.0'` is just temporary?

File: .github/workflows/docs-ci.yml
Comment (devinrsmith) [reply .github/workflows/docs-ci.yml:341]: ping

File: .github/workflows/docs-ci.yml
Comment (devinrsmith) [reply .github/workflows/docs-ci.yml:41]: We don't _need_ to have this change be seamless / atomic for deephaven.io. For example, we _could_ consider leaving `docs-ci.yml` unchanged, and create `docs-ci-v2.yml`, and placing everything under the remote_path `deephaven-core-docs-ci-v2` or something like that. It's hard for me to universally reason about what rsync may or may not do in certain circumstances. In fact, we may want to do real tests with a _different_ remote_path just to make sure it all seems to work.

File: .github/workflows/docs-ci.yml
Comment (stanbrub) [reply .github/workflows/docs-ci.yml]: That was for testing. Fixed.

File: .github/workflows/docs-ci.yml
Comment (stanbrub) [reply .github/workflows/docs-ci.yml]: Yes.  For testing.  But I should have changed it back for re-review.

**Review by devinrsmith — APPROVED** [review]
Looks good - I'm happy to merge this given the brand new directory on the remote_path. Please get approval from @dsmmcken before merge.

**Review by dsmmcken — APPROVED** [review]
Looks good, and should be low risk to merge. I'll make whatever changes we need on the other side of the fence once deployed.

**Pull request merged**
